compiler/rustc_const_eval/src/interpret/operator.rs

```diff
@@ -480,4 +481,34 @@ impl<'tcx, M: Machine<'tcx>> InterpCx<'tcx, M> {
             }
         }
     }
+
+    pub fn nullary_op(
+        &self,
+        null_op: NullOp<'tcx>,
+        arg_ty: Ty<'tcx>,
+    ) -> InterpResult<'tcx, ImmTy<'tcx, M::Provenance>> {
+        use rustc_middle::mir::NullOp::*;
+
+        let layout = self.layout_of(arg_ty)?;
+        let usize_layout = || self.layout_of(self.tcx.types.usize).unwrap();
+
+        Ok(match null_op {
+            SizeOf => {
+                assert!(layout.abi.is_sized());
+                let val = layout.size.bytes();
+                ImmTy::from_uint(val, usize_layout())
+            }
+            AlignOf => {
+                assert!(layout.abi.is_sized());
+                let val = layout.align.abi.bytes();
+                ImmTy::from_uint(val, usize_layout())
+            }
+            OffsetOf(fields) => {
+                let val =
+                    self.tcx.offset_of_subfield(self.param_env, layout, fields.iter()).bytes();
+                ImmTy::from_uint(val, usize_layout())
+            }
+            UbChecks => ImmTy::from_bool(self.tcx.sess.ub_checks(), *self.tcx),
```

Comment:
The type isn't even used here... this makes me wonder whether we should remove the type from the `Rvalue::NullaryOp` and instead have it in `NullaryOp::SizeOf` etc?

Cc @scottmcm

Reply:
I would like to see an attempt to reimplement UbChecks as a const... Somehow. It would make a lot more sense.

Reply:
> Since UbChecks is pretty weird -- it's a constant but it's not constant -- it might better to just split it out entirely since it works so differently from all the other UnOps...

It's not a constant. It can have different values in different compilation units, that's its entire point. So I think a `NullaryOp` makes a lot more sense for it than trying to make it a constant.